CNAM-286: Refactoring Patients

kevtokev opened the pull request:
Refactoring Patients :
New abstraction of Patients Extractors
Uniform all extractors of patients from different sources
Separation of extractors and filters
Separation of algorithms to calculate gender, birthDate and deathDate

dsun0720 commented on src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/patients/PatientExtractor.scala:
```diff
@@ -0,0 +1,81 @@
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.{Column, Dataset}
+
+trait PatientBIS {
```
This name is confusing. can you change a new name ?

dsun0720 commented on src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/patients/PatientExtractor.scala:
```diff
@@ -0,0 +1,81 @@
+   * @param patients that contains all patientsBIS.
+   * @return A [[Dataset]] with needed columns of Patient.
+   */
+  def fromPatientBIStoPatient(patients: Dataset[PatientType]): Dataset[Patient] = {
```
as above, I think the word PatientBis is not good enough

dsun0720 commented on src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/patients/DcirPatients.scala:
```diff
@@ -2,104 +2,170 @@
+
+import fr.polytechnique.cmap.cnam.etl.extractors.patients.PatientUtils._
+import fr.polytechnique.cmap.cnam.etl.sources.Sources
 import org.apache.spark.sql.expressions.Window
```
we use the following importing order:
java
javax
scala
all other imports
org.apache.spark
fr.polytechnique

be sure of this order in all your modified files.

dsun0720 commented on src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/patients/PatientFilters.scala:
```diff
@@ -0,0 +1,63 @@
+   * @param patients the patient to check.
+   * @return [[Dataset]] of Patient
+   */
+  def transform(patients: Dataset[Patient]): Dataset[Patient] = {
```
This is not a transform operation. I think it's better to call it, for example filterPatients

dsun0720 commented on src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/patients/PatientFilters.scala:
```diff
@@ -0,0 +1,63 @@
+   * @param maxYear the maximum year to control with.
+   * @return an Option of Patient
+   */
+  protected[patients] def controlDeathDate(minYear: Int, maxYear: Int)(patient: Patient): TraversableOnce[Patient] = {
```
This logic make me confused. In your transform method, you filter the patients not in the proper range of age but in this method you return none when a patient is not in the condition.

You need to use the same strategy.

dsun0720 commented on src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/patients/PatientFilters.scala:
```diff
@@ -0,0 +1,63 @@
+   * @param maxGender the maximum gender to control with.
+   * @return an Option of Patient
+   */
+  protected[patients] def controlGender(minGender: Int, maxGender: Int)(patient: Patient): TraversableOnce[Patient] = {
```
as above

dsun0720 commented on src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/patients/AllPatientExtractor.scala:
```diff
@@ -0,0 +1,81 @@
+import org.apache.spark.sql.functions.{coalesce, col, when}
+import org.apache.spark.sql.{Column, DataFrame, Dataset}
+
+class AllPatientExtractor {
```
better to use object instead of class, because no constructor here

dsun0720 commented on src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/patients/IrBenPatients.scala:
```diff
@@ -2,89 +2,99 @@
-      if (result.count != patients.count) {
-        throw new Exception("One or more patients have conflicting SEX CODE in IR_BEN_R")
-      }
+    val result = patients
```
why you put two distinct method here ?

dsun0720 replied:
and are you sure if it's necessary to use this kinda check here ?

dsun0720 commented on src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/patients/IrBenPatients.scala:
```diff
@@ -2,89 +2,99 @@
+    val patientids = patients.map(p => p.patientID).distinct()
 
-      result
+    if (result.count != patientids.count) {
```
are you sure if it's necessary to use this kinda check here ?

dsun0720 reviewed (changes requested):
first review

dsun0720 commented:
I looked up your new commit, i think it's ok now. But did you make a full test in the cnam to be sure that you code will produce the same result as before ?

dsun0720 commented:
add comments and squash before  merge

Commit:
CNAM-286: Refactoring Patients

New abstraction of Patients Extractors
Uniform all extractors of patients from different sources
Separation of extractors and filters
